docs/RPCS/local_intent_resolution.md

```diff
@@ -0,0 +1,59 @@
+- Start Date: 2015-07-30
+- RFC PR:
+- Cockroach Issue: #1821 (now obsolete)
+
+# Summary
+
+Asynchronous intent resolution upon Transaction completion should be moved from client gateway to `Range`. An `EndTransaction` call should contain a definitive list of intents belonging to the `Transaction`, and the `Range` executing the `EndTransaction` call (i.e. the base `Range` for the `Transaction`) can in most cases resolve most (if not all) intents synchronously and often garbage-collect the record right away. In all other cases, the Range will perform the asynchronous resolution of the remaining intents, and a `TransactionGCQueue` can periodically scrub those `Transaction`'s records. 
+
+# Motivation
+
+Currently, the client gateways (`TxnCoordSender`) involved in executing a Transaction (`Txn`) keep individual records of the intents written throughout the `Txn`. Upon completion (that is, when `EndTransaction` either commits or aborts the `Txn`), the client gateway executing that command asynchronously resolves the intents tracked by it. There are some issues with that:
+
+* intent resolution happens fairly late and is a few round-trip latencies away from the most likely intent location: The range which holds the `Txn`'s record.
+* a `Txn` carried out over multiple coordinators will not have all of its intents resolved, only those on the coordinator which executes `EndTransaction`.
+* partially due to the above, it's hard to garbage-collect `Txn` records: it's not clear whether there are any open intents somewhere else in the system.
+
+Not resolving intents is not a correctness issue, but can significantly reduce performance in the presence of contention as conflict resolution requires a lot of round-trip delays and potential backoff. Not garbage-collecting `Txn` records is also not a correctness issue, but necessary at some point.
+
+At the Range level, we can deal with most of these issues, and adding a bit more data to the `Transaction` records which can not be gc'ed immediately allows speedy cleanup of the rarer cases as well.
+
+# Detailed design
+
+## Disallow cross-coordinator Transactions
+Dealing with multiple coordinators handling a single `Txn` is tricky: It's hard to keep clients from using more than one gateway reliably without additional complexity, but the natural path to take with the suggested changes is to declare all intents which are not contained in the transaction record illegal. This could mean that parts of a client's `Transaction` might get lost upon commit when written through another gateway, unless we take precautions.
+
+An easy way out is adding a sequence number to the `Transaction` object whose single purpose is to be incremented by the client before sending another request in a `Txn`: The gateway aborts the `Transaction` if the sequence number does not match; additionally, a Transaction can only start with sequence number one. The returned error indicates an abort of the `Txn` with immediate retry.
```

Comment:
We can disallow cross-coordinator transactions without adding a new field. The first request in a transaction is already special since it does not include a transaction ID. We could move the creation of the heartbeat goroutine from `sendOne` to `maybeBeginTxn` and return an error if we receive a txn ID that is not in our `txns` map.

If the sequence number is useful for other purposes then it may be worth including, but I'd like to avoid it if we can because it removes opportunities for parallelism (with a sequence number the only way to parallelize operations is to include them in the same batch; without a sequence number requests may be sent separately in parallel as long as the client is smart enough to merge all the transaction objects before calling `EndTransaction`).